File: ios/Shared/Sources/Core/Models/LmsRequestLogEntry.swift
import Foundation

public struct LmsRequestLogContext: Codable, Equatable, Sendable {
    public let navigationPath: String
    public let pageTitle: String

    public init(navigationPath: String, pageTitle: String) {
        self.navigationPath = navigationPath
        self.pageTitle = pageTitle
    }
}

public struct LmsRequestLogEntry: Codable, Identifiable, Equatable, Sendable {
    public struct Parameter: Codable, Equatable, Sendable {
        public let name: String
        public let value: String?

        public init(name: String, value: String?) {
            self.name = name
            self.value = value
        }

        public var displayText: String {
            if let value {
                return "\(name)=\(value)"
            }

            return "\(name)=<nil>"
        }
    }

    public let id: UUID
    public let occurredAt: Date
    public let context: LmsRequestLogContext
    public let wsFunction: String
    public let parameters: [Parameter]
    public let statusCode: Int?
    public let errorMessage: String?
    public let responseBody: String

    public init(
        id: UUID = UUID(),
        occurredAt: Date = .now,
        context: LmsRequestLogContext,
        wsFunction: String,
        parameters: [Parameter],
        statusCode: Int?,
        errorMessage: String?,
        responseBody: String
    ) {
        self.id = id
        self.occurredAt = occurredAt
        self.context = context
        self.wsFunction = wsFunction
        self.parameters = parameters
        self.statusCode = statusCode
        self.errorMessage = errorMessage
        self.responseBody = responseBody
    }

    public var hasError: Bool {
        errorMessage != nil || statusCode.map { 200..<300 ~= $0 } == false
    }

    public var metadataText: String {
        var lines = [
            "path: \(context.navigationPath)",
            "title: \(context.pageTitle)",
            "wsfunction: \(wsFunction)",
        ]

        if parameters.isEmpty == false {
            lines.append("parameters:")
            lines.append(contentsOf: parameters.map { "  - \($0.displayText)" })
        }

        if let statusCode {
            lines.append("status: \(statusCode)")
        }

        if let errorMessage, errorMessage.isEmpty == false {
            lines.append("error: \(errorMessage)")
        }

        return lines.joined(separator: "\n")
    }

    public var displayText: String {
        if responseBody.isEmpty {
            return "\(metadataText)\n\n<empty response>"
        }

        return "\(metadataText)\n\n\(responseBody)"
    }
}

File: android/core/src/main/kotlin/dev/example/moodlenative/core/LmsRequestLogEntry.kt
package dev.example.moodlenative.core

import java.time.Instant
import java.util.UUID

data class LmsRequestLogContext(
    val navigationPath: String,
    val pageTitle: String,
)

data class LmsRequestLogEntry(
    val id: UUID = UUID.randomUUID(),
    val occurredAt: Instant = Instant.now(),
    val context: LmsRequestLogContext,
    val wsFunction: String,
    val parameters: List<Parameter>,
    val statusCode: Int?,
    val errorMessage: String?,
    val responseBody: String,
) {
    data class Parameter(
        val name: String,
        val value: String?,
    ) {
        val displayText: String
            get() = if (value != null) "$name=$value" else "$name=<nil>"
    }

    val hasError: Boolean
        get() = errorMessage != null || statusCode?.let { it !in 200..299 } == true

    val metadataText: String
        get() = buildList {
            add("path: ${context.navigationPath}")
            add("title: ${context.pageTitle}")
            add("wsfunction: $wsFunction")
            if (parameters.isNotEmpty()) {
                add("parameters:")
                addAll(parameters.map { "  - ${it.displayText}" })
            }
            if (statusCode != null) {
                add("status: $statusCode")
            }
            if (!errorMessage.isNullOrEmpty()) {
                add("error: $errorMessage")
            }
        }.joinToString(separator = "\n")

    val displayText: String
        get() = if (responseBody.isEmpty()) {
            "$metadataText\n\n<empty response>"
        } else {
            "$metadataText\n\n$responseBody"
        }
}
